torch/csrc/autograd/engine.cpp

```diff
@@ -303,7 +303,9 @@ auto Engine::thread_init(int device) -> void {
 // completely unrelated to that of Eval1 (it's not a recursive call).
 // It's all ok and is handled right now, but it should be accounted for
 // in case this code is to be changed.
-auto Engine::thread_main(GraphTask *graph_task) -> void {
+auto Engine::thread_main(std::shared_ptr<GraphTask> graph_task) -> void {
+  auto device_lock = device_locks_[worker_device+1];
```

Comment:
Taking an owning reference so mutex doesn't get destroyed while we still have it taken out in a thread!